mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java

```diff
@@ -415,6 +418,21 @@ private static Object prepareIdObjectValue(String[] columnNames, Object[] column
 		return getCollection( key.getTable() );
 	}
 
+	private MongoCollection<Document> withOptions(MongoCollection<Document> collection,  ReadPreference readPreference, ReadConcern readConcern) {
+		if ( readPreference != null && readConcern != null ) {
```

Comment:
I think this method should also apply the writeConcern and should be used everytime with need a collection.
This way the options are applied always by the same function.

I also find this code harder to read than needed, what do you think about refactoring it to look something like:

```
	private MongoCollection<Document> withOptions(MongoCollection<Document> collection, AssociationContext context) {
		return withOptions( collection, context.getAssociationTypeContext().getOptionsContext() );
	}

	private MongoCollection<Document> withOptions(MongoCollection<Document> collection, TupleTypeContext context) {
		return withOptions( collection, context.getOptionsContext() );
	}

	private MongoCollection<Document> withOptions(MongoCollection<Document> collection, OptionsContext context) {
		MongoCollection<Document> newCollection = collection;
		newCollection = withReadPreference( context, newCollection );
		newCollection = withReadConcern( context, newCollection );
		return newCollection;
	}

	private MongoCollection<Document> withReadConcern(OptionsContext context, MongoCollection<Document> newCollection) {
		ReadConcern readConcern = context.getUnique( ReadConcernOption.class );
		if ( readConcern != null ) {
			newCollection = newCollection.withReadConcern( readConcern );
		}
		return newCollection;
	}

	private MongoCollection<Document> withReadPreference(OptionsContext context, MongoCollection<Document> newCollection) {
		ReadPreference readPreference = context.getUnique( ReadPreferenceOption.class );
		if ( readPreference != null ) {
			return newCollection.withReadPreference( readPreference );
		}
		return newCollection;
	}
```

Now that think about it it might make sense to call withOptions everytime we call the getCollection( ),
something like:

```
	private MongoCollection<Document> getCollection(String table, ...) {
		return withOptions( currentDB.getCollection( table ), ... );
	}
```

Reply:
> for example the classes and the tests could be in separate commits.

What I meant is the *new* classes and the new tests could be in their own commits